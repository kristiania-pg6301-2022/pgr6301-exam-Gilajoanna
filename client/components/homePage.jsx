import { fetchJSON } from "../globals/fetchJSON";
import { ListArticles } from "./articles";
import * as React from "react";

export function HomePage({ user }) {
  async function listAllArticles() {
    return await fetchJSON("/api/articles");
  }

  if (!user) {
    return <div>Please log in to have full access to articles.</div>;
  }

  return (
    <>
      <div>
        <article>
          <ListArticles listAllArticles={listAllArticles} />
        </article>
      </div>
    </>
  );
}
