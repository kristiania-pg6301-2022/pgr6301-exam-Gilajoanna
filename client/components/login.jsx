import * as React from "react";
import { Route, Routes, useNavigate } from "react-router-dom";
import { useContext, useEffect, useState } from "react";
import { AppContext } from "../globals/AppContext.jsx";

function Login({ config }) {
  return (
    <div>
      <h2>Login to read articles</h2>
      <LoginButton label={"Login with Google"} config={config} />
    </div>
  );
}

function LoginButton({ config, label }) {
  async function handleLogin() {
    const { authorization_endpoint, response_type, scope, client_id } = config;
    const params = {
      response_type,
      client_id,
      scope,
      redirect_uri: `${window.location.origin}/login/callback`,
    };

    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(params);
  }

  return (
    <div>
      <button onClick={handleLogin}>{label}</button>
    </div>
  );
}

function LoginCallback({ reload }) {
  const [error, setError] = useState();
  const navigate = useNavigate();
  const { registerLogin } = useContext(AppContext);

  useEffect(async () => {
    const { access_token } = Object.fromEntries(
      new URLSearchParams(window.location.hash.substring(1))
    );

    if (access_token) {
      const response = await fetch("/api/login", {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({ access_token }),
      });

      if (!response.ok) {
        setError(
          `Failed to fetch token ${response.status}: ${await response.text()}`
        );
        return;
      }
      await registerLogin({ access_token });
      reload();
      navigate("/");
    }
  });

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>Something went wrong. Error: {error}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>Please wait...</h1>
    </div>
  );
}

function LogOut({ reload }) {
  const navigate = useNavigate();
  const { registerLogout } = useContext(AppContext);
  useEffect(async () => {
    await registerLogout();
    reload();
    navigate("/");
  });

  return <h1>Please wait...</h1>;
}

export function LoginPage({ config, reload }) {
  return (
    <Routes>
      <Route path={""} element={<Login config={config} />} />
      <Route path={"/callback"} element={<LoginCallback reload={reload} />} />
      <Route path={"/logout"} element={<LogOut reload={reload} />} />
      <Route path={"*"} element={<Login config={config} />} />
    </Routes>
  );
}
