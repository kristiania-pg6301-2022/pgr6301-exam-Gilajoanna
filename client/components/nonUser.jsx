import { fetchJSON } from "../globals/fetchJSON";
import { AsideArticleList } from "./articles";
import * as React from "react";

export function NonUser() {
  async function listAllArticles() {
    return await fetchJSON("/api/articles");
  }

  return (
    <article>
      <AsideArticleList listAllArticles={listAllArticles} />
    </article>
  );
}
