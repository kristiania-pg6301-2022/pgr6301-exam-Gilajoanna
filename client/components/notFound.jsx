import { Link } from "react-router-dom";
import * as React from "react";

export function NotFound() {
  return (
    <div>
      <h1>Page not found</h1>
      <div>
        <Link to={"/"}>Back to home page</Link>
      </div>
    </div>
  );
}
