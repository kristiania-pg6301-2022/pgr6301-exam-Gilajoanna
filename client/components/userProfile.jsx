import { useLoader } from "../globals/useLoader";
import { fetchJSON } from "../globals/fetchJSON";
import * as React from "react";

export function UserProfile({ user }) {
  const { loading, error } = useLoader(async () => {
    return await fetchJSON("/api/login");
  });

  if (loading) {
    return <div>Loading..</div>;
  }

  if (error) {
    return <div id={"error-message"}>Error! {error.toString()}</div>;
  }

  if (!user) {
    return <div>You must be logged in to access your profile.</div>;
  }

  return (
    <div>
      <h1>
        Profile for {user.name} ({user.email}){" "}
      </h1>
      <div>
        <img src={user.picture} alt={"Profile image"} />
      </div>
    </div>
  );
}
