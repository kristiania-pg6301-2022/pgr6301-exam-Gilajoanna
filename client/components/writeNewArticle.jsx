import * as React from "react";
import { useContext, useState } from "react";
import { AppContext } from "../globals/AppContext";

function ArticleForm({ label, value, onChangeValue }) {
  return (
    <div>
      <label>{label}</label>{" "}
      <input
        type={"text"}
        name={"title"}
        value={value}
        onChange={(e) => onChangeValue(e.target.value)}
      />
    </div>
  );
}

export function WriteNewArticle() {
  const { postArticle } = useContext(AppContext);

  const [title, setTitle] = useState("");
  const [category, setCategory] = useState("");
  const [content, setContent] = useState("");
  const [author, setAuthor] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();

    if (!title || !category || !author || !content) {
      alert("One or more inputs are empty. Please try again.");
    }

    await postArticle({ title, category, author, content });

    setTitle("");
    setCategory("");
    setAuthor("");
    setContent("");
  }

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <h1>Write new article</h1>

        <ArticleForm label={"Title:"} value={title} onChangeValue={setTitle} />
        <ArticleForm
          label={"Category:"}
          value={category}
          onChangeValue={setCategory}
        />
        <ArticleForm
          label={"Content:"}
          value={content}
          onChangeValue={setContent}
        />
        <ArticleForm
          label={"Author:"}
          value={author}
          onChangeValue={setAuthor}
        />
        <button type={"submit"}>Publish</button>
      </form>
    </div>
  );
}
