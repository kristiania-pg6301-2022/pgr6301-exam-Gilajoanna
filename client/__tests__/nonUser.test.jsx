import * as ReactDOM from "react-dom";
import * as React from "react";
import { NonUser } from "../components/nonUser";

describe("Non user component", () => {
  it("renders Non user", () => {
    const element = document.createElement("div");
    ReactDOM.render(<NonUser />, element);
    expect(element.innerHTML).toMatchSnapshot();
  });
});
