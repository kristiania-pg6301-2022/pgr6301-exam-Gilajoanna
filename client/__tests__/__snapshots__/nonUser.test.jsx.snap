// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Non user component renders Non user 1`] = `"<article><div>Please wait..</div></article>"`;
